### PR TITLE
CMake: Update build system to default to use Ninja

### PR DESCRIPTION
**Pull request by kwryankrattiger**
Ninja is pretty much ubiquitous across all system now, as well as being cross platform (for future Windows support) and significantly faster than `Unix Makefiles` . Unless there is something about it I am not aware of, it should be safe to turn on as the default.

### CODE REVIEW

File: lib/spack/docs/tables/system_prerequisites.csv [hunk 1]
```diff
@@ -2,6 +2,7 @@ Name, Supported Versions, Notes, Requirement Reason
 Python, 2.7/3.5-3.9, , Interpreter for Spack
 C/C++ Compilers, , , Building software
 make, , , Build software
+ninja, , , Build software
```
Comment (adamjstewart) [inline lib/spack/docs/tables/system_prerequisites.csv:5]: This isn't a system prereq, Spack installs it itself

File: lib/spack/spack/build_systems/cmake.py [hunk 2]
```diff
@@ -88,7 +88,8 @@ class CMakePackage(PackageBase):
     #:
     #: See https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html
     #: for more information.
-    generator = 'Unix Makefiles'
+    generator = 'Ninja'
+    depends_on('ninja')
```
Comment (adamjstewart) [inline lib/spack/spack/build_systems/cmake.py:92]:
```suggestion
    depends_on('ninja', type='build')
```

File: var/spack/repos/builtin.mock/packages/ninja/package.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+class Ninja(Package):
```
Comment (adamjstewart) [inline var/spack/repos/builtin.mock/packages/ninja/package.py:7]: This can probably just be a minimal package, we don't need to copy the entire thing

**Review by adamjstewart — COMMENTED** [review]
I have heard that there are some packages that don't work with ninja, although I haven't personally encountered this.

**Comment by jrood-nrel** [comment]
I was an advocate of this early on, but I am fully against it now. If there is _any_ Fortran in your application, Ninja will burn you. It's supposed to have Fortran support thanks to Kitware, but I promise you will have problems with mixed-Fortran applications. Without Fortran it's great.

File: lib/spack/docs/tables/system_prerequisites.csv (hunk 1 above)
Comment (kwryankrattiger) [reply lib/spack/docs/tables/system_prerequisites.csv:5]: I wasn't sure if it was required for bootstrapping, good to know.

**Comment by kwryankrattiger** [comment]
@jrood-nrel 
Ninja has officially supported Fortan since [1.10](https://groups.google.com/g/ninja-build/c/piOltAhywFA/m/zPfkrTtRCwAJ) which released in 2020. In my experience mixed language projects have been working fine since then. Doing testing on the mixed language packages in the ecp-data-vis-sdk and I have had no issues with using ninja to build.

The biggest issue I am finding is `make()` being used everywhere in `CMakePackage`s directly. It seems a lot of packages have these issues because they inherit `CMakePackage` and then don't use any of the framework it provides. The main reason for this seems to be packages want to install the CMakeCache.txt file along with the project, which I think if that is something people want would make sense to just put into the `CMakePackage.install` phase by default anyway.

But just gauging some basic tests this PR needs a lot of other fixes in order to work anyway. I have stuff to fix the dray/conduit/ascent/vtk-h issues but that is the only stack I addressed. They aren't really fixes either, just re-hardcoding the packages for a different default build tool which isn't really a solution.

If this patch were to be merge-able it will require fixes to many projects and I don't really have the time to do that, so I am going to close this PR and punt all Ninja fixes to #27021 and the subsequent changes packages will need to make in order to support Windows.

**Comment by jrood-nrel** [comment]
Thanks for the update. We have used ninja extensively in my project which has a huge stack (`nalu-wind`) and while we can succeed for the most part with several mixed fortran/c++ applications in our stack. Problems have manifested in different ways on different platforms, until ultimately we had to give up enabling ninja locally in all our custom packages. I wish we didn't have to give up because building `trilinos` alone saves us 20% in compile time with ninja.

**Comment by jrood-nrel** [comment]
I know this is closed, but for future reference on why Ninja continues to not be ready for every project, I noticed this example https://github.com/spack/spack/pull/29163

**Pull request closed without merging**